File: CorteReal.py
import sys

# ************************************************
#   CSC 1800-002 Programming Assignment #2
#   Last updated by Miguel Corte-Real on 9/12/16
# ************************************************

# Initialize a dictionary representing the family tree
family_tree = {}


# Functions used to handle the user input
def handle_event(splitstr):
    p1 = get_person(splitstr[1])
    p2 = get_person(splitstr[2])
    if not are_married(p1, p2):
        p1.spouses.append(p2)
        p2.spouses.append(p1)
    if len(splitstr) is 4:
        p3 = Person(splitstr[3], p1, p2)
        family_tree[p3.name] = p3
        p1.children.append(p3)
        p2.children.append(p3)


def handle_r_query(splitstr):
    p1 = get_person(splitstr[1])
    p2 = get_person(splitstr[2])
    if p1 in p2.spouses:
        return "spouse"
    if p1 in p2.parents:
        return "parent"
    if p1 in get_siblings(p2):
        return "sibling"
    if p1 in get_half_siblings(p2):
        return "half-sibling"
    # TODO: ancestor
    # TODO: cousin
    return "unrelated"


def handle_x_query(splitstr):
    p1 = get_person(splitstr[1])
    p2 = get_person(splitstr[3])
    relation = splitstr[2]
    if relation == "spouse":
        return p1 in p2.spouses
    elif relation == "parent":
        return p1 in p2.parents
    elif relation == "sibling":
        return p1 in get_siblings(p2)
    elif relation == "half-sibling":
        return p1 in get_half_siblings(p2)
    # TODO: ancestor
    # TODO: cousin
    # TODO: unrelated


def handle_w_query(splitstr):
    p = get_person(splitstr[2])
    relation = splitstr[1]
    names = []
    if relation == "spouse":
        names = sorted([s.name for s in p.spouses])
    if relation == "parent":
        names = sorted([r.name for r in p.parents])
    if relation == "sibling":
        names = sorted([s.name for s in get_siblings(p)])
    if relation == "half-sibling":
        names = sorted([hs.name for hs in get_half_siblings(p)])
    # TODO: ancestor
    # TODO: cousin
    # TODO: unrelated
    for name in names:
        print(name)


def are_married(p1, p2):
    return p2 in p1.spouses and p1 in p2.spouses


def get_siblings(p):
    if len(p.parents) == 0:
        return []
    siblings = list(set(p.parents[0].children) & set(p.parents[1].children))
    return [s for s in siblings if s is not p]


def get_half_siblings(p):
    if len(p.parents) == 0:
        return[]
    half_siblings = list((set(p.parents[0].children) | set(p.parents[1].children))
                         - (set(p.parents[0].children) & set(p.parents[1].children)))
    return [hs for hs in half_siblings if hs is not p]

def get_set_of_ancestors(p):
    # Here is essentially my code from java
    # It is not possible to use a recursive solution and not include the parents
    # So this is going to be called from the actual get_ancestors function for each parent
    # This is kinda dirty but I think it's cleaner and less confusing than doing it without recursion
    ancestorSet = set(p.parents)
    
    if len(p.parents) == 0:
        return ancestorSet
    
    return ancestorSet | get_set_of_ancestors(p.parents[0]) | get_set_of_ancestors(p.parents[1])


def get_ancestors(p):
    # All this does is call the actual method on each parent and combine them
    if len(p.parents) == 0:
        return set([])
    
    return get_set_of_ancestors(p.parents[0]) | get_set_of_ancestors(p.parents[1])

def get_person(name):
    if name in family_tree.keys():
        person = family_tree.get(name)
    else:
        person = Person(name)
        family_tree[person.name] = person
    return person


# The Person class
class Person:
    def __init__(self, name, parent1=None, parent2=None):
        self.name = name
        self.parents = []
        if parent1 is not None:
            self.parents.insert(0, parent1)
        if parent2 is not None:
            self.parents.insert(1, parent2)
        self.children = []
        self.spouses = []

    def add_child(self, child):
        self.children.append(child)

    def add_spouse(self, spouse):
        if spouse not in self.spouses:
            self.spouses.append(spouse)


# Read the input, build tree, and handle queries
for line in sys.stdin:
    delim_line = line.rstrip().split(' ')
    if line.startswith('E'):
        handle_event(delim_line)
    else:
        print("\n" + line.rstrip())
        if line.startswith('R'):
            print(handle_r_query(delim_line))
        elif line.startswith('X'):
            print(handle_x_query(delim_line))
        elif line.startswith('W'):
            handle_w_query(delim_line)

File: javaFiles/CorteReal.java
import java.io.FileInputStream;
import java.io.IOException;
import java.util.*;

/**
 * CSC 1800-002 Programming Assignment #1
 * Last updated by Miguel Corte-Real on 9/11/16.
 */

public class CorteReal {
    public static void main (String[] args) throws IOException {
        Scanner fileInput = new Scanner(System.in);

        HashMap<String, Person> family = new HashMap<>();

        while (fileInput.hasNextLine()) {
            String currLine = fileInput.nextLine();
            Scanner lineInput = new Scanner(currLine);
            String lineType = lineInput.next();
            switch (lineType) {
                case "E" :
                    handleEvent(lineInput.nextLine(), family);
                    break;
                case "R" :
                    System.out.println("\n" + currLine);
                    System.out.println(handleRQuery(lineInput.nextLine(), family));
                    break;
                case "X" :
                    System.out.println("\n" + currLine);
                    System.out.println(handleXQuery(lineInput.nextLine(), family));
                    break;
                case "W" :
                    System.out.println("\n" + currLine);
                    handleWQuery(lineInput.nextLine(), family).forEach(System.out::println);
                    break;
            }
        }
    }

    private static void handleEvent(String people, HashMap<String, Person> family) {
        String[] names = people.split(" ");

        Person person1 = family.get(names[1]);
        if (person1 == null) {
            person1 = new Person(names[1]);
            family.put(names[1], person1);
        }

        Person person2 = family.get(names[2]);
        if (person2 == null) {
            person2 = new Person(names[2]);
            family.put(names[2], person2);
        }

        person1.optionalAddSpouse(person2);
        person2.optionalAddSpouse(person1);

        Person person3;
        if (names.length == 4) {
            person3 = family.get(names[3]);
            if (person3 == null) {
                person3 = new Person(names[3], person1, person2);
                family.put(names[3], person3);
            }
            person1.addChild(person3);
            person2.addChild(person3);
        }
    }

    private static String handleRQuery(String query, HashMap family) {
        String[] queryItems = query.split(" ");

        Person p1 = (Person)family.get(queryItems[1]);
        Person p2 = (Person)family.get(queryItems[2]);
        if (p1 == null || p2 == null) return "unrelated";

        Set<Person> relations = new HashSet<>();

        if (p2.getSpouses().contains(p1)) return "spouse";
        if (p2.getParents().contains(p1)) return "parent";
        if (p2.getSiblings().contains(p1)) return "sibling";
        if (p2.getHalfSiblings().contains(p1)) return "half-siblings";
        if (p2.getAncestors(relations).contains(p1)) return "ancestor";
        if (p2.getCousins().contains(p1)) return "cousin";
        return "unrelated";
    }

    private static boolean handleXQuery(String query, HashMap<String, Person> family) {
        String[] queryItems = query.split(" ");

        Person p1 = family.get(queryItems[1]);
        Person p2 = family.get(queryItems[3]);
        if (p1 == null || p2 == null) return false;

        Set<Person> relations = new HashSet<>();

        switch (queryItems[2]) {
            case "spouse" :
                return (p2.getSpouses().contains(p1));
            case "parent" :
                return (p2.getParents().contains(p1));
            case "sibling" :
                return (p2.getSiblings().contains(p1));
            case "half-sibling" :
                return (p2.getHalfSiblings().contains(p1));
            case "ancestor" :
                return (p2.getAncestors(relations).contains(p1));
            case "cousin" :
                return (p2.getCousins().contains(p1));
            case "unrelated" :
                return (getUnrelated(p2, family).contains(p1));
        }
        return false;
    }

    private static List<Person> handleWQuery(String query, HashMap<String, Person> family) {
        String[] queryItems = query.split(" ");

        Person p = family.get(queryItems[2]);
        if (p == null) return new ArrayList<>();

        Set<Person> relations = new HashSet<>();

        switch (queryItems[1]) {
            case "spouse" :
                List<Person> sortedSpouses = p.getSpouses();
                Collections.sort(sortedSpouses);
                return sortedSpouses;
            case "parent" :
                List<Person> sortedParents = p.getParents();
                Collections.sort(sortedParents);
                return sortedParents;
            case "sibling" :
                relations = p.getSiblings();
                break;
            case "half-sibling" :
                relations = p.getHalfSiblings();
                break;
            case "ancestor" :
                relations = p.getAncestors(relations);
                relations.remove(p);
                break;
            case "cousin" :
                relations = p.getCousins();
                break;
            case "unrelated" :
                relations = getUnrelated(p, family);
        }
        List<Person> sortedRelations = new ArrayList<>(relations);
        Collections.sort(sortedRelations);
        return sortedRelations;
    }

    private static Set<Person> getUnrelated(Person p, HashMap<String, Person> family) {
        Set<Person> relations = new HashSet<>();
        relations = p.getAncestors(relations);
        relations.addAll(p.getSpouses());
        relations.addAll(p.getSiblings());
        relations.addAll(p.getHalfSiblings());
        relations.addAll(p.getCousins());

        Set<Person> unrelated = new HashSet<>();
        unrelated.addAll(family.values());
        unrelated.removeAll(relations);

        return unrelated;
    }

    private static class Person implements Comparable {
        String name;
        Person[] parents = new Person[2];
        List<Person> spouses;
        List<Person> children;

        private Person(String name) {
            this.name = name;
            spouses = new ArrayList<>();
            children = new ArrayList<>();
        }

        private Person(String name, Person parent1, Person parent2) {
            this.name = name;
            parents[0] = parent1;
            parents[1] = parent2;
            spouses = new ArrayList<>();
            children = new ArrayList<>();
        }

        private void optionalAddSpouse(Person spouse) {
            if (!spouses.contains(spouse)) {
                spouses.add(spouse);
            }
        }

        private void addChild(Person child) {
            children.add(child);
        }

        private Set<Person> getSiblings() {
            Set<Person> p1Children = new HashSet<>();
            Set<Person> p2Children = new HashSet<>();
            if (parents[0] != null) p1Children.addAll(parents[0].getChildren());
            if (parents[1] != null) p2Children.addAll(parents[1].getChildren());
            p1Children.retainAll(p2Children);
            p1Children.remove(this);
            return p1Children;
        }

        private Set<Person> getHalfSiblings() {
            Set<Person> p1Children = new HashSet<>();
            Set<Person> p2Children = new HashSet<>();
            Set<Person> intersection = new HashSet<>();
            Set<Person> tmp = new HashSet<>();

            if (parents[0] != null) p1Children.addAll(parents[0].getChildren());
            if (parents[1] != null) p2Children.addAll(parents[1].getChildren());

            intersection.addAll(p1Children);
            intersection.retainAll(p2Children);

            tmp.addAll(p1Children);
            tmp.addAll(p2Children);

            tmp.removeAll(intersection);
            return tmp;
        }

        private Set<Person> getAncestors(Set<Person> ancestors) {
            ancestors.add(this);
            if (parents[0] != null) {
                ancestors.addAll(parents[0].getAncestors(ancestors));
                ancestors.addAll(parents[1].getAncestors(ancestors));
            }
            return ancestors;
        }

        private Set<Person> getCousins() {
            Set<Person> closeCousins = new HashSet<>();
            Set<Person> visitedCousins = new HashSet<>();
            Set<Person> ancestors = new HashSet<>();
            Set<Person> cousins = new HashSet<>();

            this.getAncestors(ancestors).forEach(ancestor -> closeCousins.addAll(ancestor.getChildren()));

            for (Person cousin : closeCousins) {
                if (!visitedCousins.contains(cousin)) cousins.addAll(cousin.getChildren());
                visitedCousins.add(cousin);
            }

            cousins.addAll(closeCousins);
            cousins.removeAll(getParents());
            cousins.removeAll(getSiblings());
            cousins.removeAll(getHalfSiblings());
            cousins.remove(this);
            return cousins;
        }

        private String getName() {
            return name;
        }

        private List<Person> getSpouses() {
            return spouses;
        }

        private List<Person> getChildren() {
            return children;
        }

        private List<Person> getParents() {
            if (parents[0] != null) {
                return Arrays.asList(parents);
            }
            return new ArrayList<>();
        }

        @Override
        public int compareTo(Object o) {
            Person p = (Person)o;
            return name.compareTo(p.getName());
        }

        @Override
        public String toString() {
            return name;
        }
    }
}
